app/src/org/commcare/activities/connect/ConnectIdDatabaseHelper.java

```diff
@@ -136,4 +140,74 @@ public static void setRegistrationPhase(Context context, ConnectIdTask phase) {
             storeUser(context, user);
         }
     }
+
+    public static void storeAvailableJobs(Context context, List<ConnectJob> jobs) {
+        List<ConnectJob> existingList = getAvailableJobs(context);
+
+        //Delete jobs that are no longer available
+        for (ConnectJob existing : existingList) {
+            boolean stillExists = false;
+            for (ConnectJob newJob : jobs) {
+                if(existing.getID() == newJob.getID()) {
+                    stillExists = true;
+                    break;
+                }
+            }
+
+            if(!stillExists) {
+                getConnectStorage(context, ConnectJob.class).remove(existing.getID());
+            }
+        }
+
+        //Now insert/update jobs
+        for (ConnectJob newJob : jobs) {
+            for (ConnectJob existingJob : existingList) {
+                if(newJob.getID() == existingJob.getID()) {
+                    //To update, set ID for the new job from the existing
+                    newJob.setID(existingJob.getID());
+                    break;
+                }
+            }
```

Comment:
can we merge this with the loop to find non-existing jobs ? They are both nXn loops so I will prefer to only do these iterations once if possible.

Reply:
Improved in [f6c3b16](https://github.com/dimagi/commcare-android/pull/2695/commits/f6c3b1643a87250b1f4fc7627167691b72f4c1c) so that the first nXn loop sets the ID for jobs found in the existing list, reducing the second nXn loop to only n.